cassovary-core/src/main/scala/com/twitter/cassovary/graph/MemoryMappedDirectedGraph.scala

```diff
@@ -122,190 +159,164 @@ object MemoryMappedDirectedGraph {
+      val tempFiles = neighborType match {
+        case GraphDir.OutDir => tempFilesById1
+        case GraphDir.InDir => tempFilesById2
+      }
+      for ((tempFile, chunkIndex) <- tempFiles.zipWithIndex) {
+        // Read edges from temporary file into arrays
+        val inStream = new DataInputStream(new BufferedInputStream(
+          new FileInputStream(tempFile)))
+
+        // The last chunk might not have the full number of nodes, so compute # nodes in this chunk.
+        val nodeCountInChunk = math.min(nodesPerChunk, nodeCount - chunkIndex * nodesPerChunk)
+        val neighborArrays = Array.fill(nodeCountInChunk)(new IntArrayList())
+        // The difference between a nodeId and the corresponding index into neighborArrays
+        val nodeOffset = chunkIndex * nodesPerChunk
+        val tempFileEdgeCount = tempFile.length() / 8L
+        for (edgeIndex <- 0L until tempFileEdgeCount) {
+          val id1 = inStream.readInt()
+          val id2 = inStream.readInt()
+          neighborType match {
+            case GraphDir.OutDir => neighborArrays(id1 - nodeOffset).add(id2)
+            case GraphDir.InDir => neighborArrays(id2 - nodeOffset).add(id1)
+          }
+        }
+        inStream.close()
+
+        // Write edge data to binary file and store neighbor offsets
+        val neighborOffsets = new Array[Long](nodeCountInChunk)
+        for ((neighbors, i) <- neighborArrays.zipWithIndex) {
+          neighborOffsets(i) = cumulativeNeighborOffset
+          val sortedDistinctNeighbors = FastUtilUtils.sortedDistinctInts(neighbors)
+          for (j <- 0 until sortedDistinctNeighbors.size()) {
+            val neighborId = sortedDistinctNeighbors.get(j)
+            binaryGraphOutput.writeInt(neighborId)
+            halfEdgeCount += 1
+            if (halfEdgeCount % (100 * 1000 * 1000) == 0)
+              log(s"wrote $halfEdgeCount half edges")
+            cumulativeNeighborOffset += 4
+          }
+        }
+        // Store the ending offset for the last node
+        val finalOffset = cumulativeNeighborOffset
+        binaryGraphOutput.flush()
+
+        // Write neighbor offsets to binary file
+        val chunkOffsetsStart = neighborType match {
+          case GraphDir.OutDir => 16L + 8L * chunkIndex * nodesPerChunk
+          case GraphDir.InDir => 16L + 8L * chunkIndex * nodesPerChunk + 8L * (nodeCount + 1)
+        }
+        val offsetsBuffer = binaryGraphChannel.map(MapMode.READ_WRITE, chunkOffsetsStart,
+          8 * (nodeCountInChunk + 1)) // The + 1 is for the edge case described below
+
+        for (offset <- neighborOffsets) {
+          offsetsBuffer.putLong(offset)
+        }
+        // Edge case: For the last chunk, we need to write the final offset, in order to store
+        // the nth node's out-degree (and in-degree).
+        if (nodesPerChunk * (chunkIndex + 1) >= nodeCount) {
+          offsetsBuffer.putLong(finalOffset)
+        }
+        offsetsBuffer.force()
```

Comment:
Yes. Both copy data from RAM to disk.  I'll add a comment.

 I'm not sure how to cleanly get around Java's lack of a buffered random access file. I think the right way would be to write our own class which wraps a file, does the buffering, and allows seeking and writing ints and longs.  It probably wouldn't be worth the effort though.